app/models/game.rb

```diff
@@ -34,6 +35,47 @@ def determine_check
     false
   end
 
+  def determine_checkmate
+    return white_checkmate if turn_number.even? && determine_check
+    black_checkmate if turn_number.odd? && determine_check
+  end
+
+  def black_checkmate
+    black_pieces = pieces.where(user_id: black_player_id)
+    check_status = true
+    black_pieces.each do |piece|
+      8.times do |row|
+        8.times do |col|
+          next if piece.valid_move?(row, col)
+          Piece.transaction do
+            piece.move_to!(row, col)
+            check_status = false if determine_check == false
+            fail ActiveRecord::Rollback
+          end
+        end
+      end
+    end
+    check_status
+  end
+
+  def white_checkmate
+    white_pieces = pieces.where(user_id: white_player_id)
+    check_status = true
+    white_pieces.each do |piece|
+      8.times do |row|
+        8.times do |col|
+          next if piece.valid_move?(row, col)
+          Piece.transaction do
+            piece.move_to!(row, col)
+            check_status = false if determine_check == false
+            fail ActiveRecord::Rollback
+          end
+        end
+      end
+    end
+    check_status
+  end
+
```

Comment:
They're absolutely identical.  That's a fantastic idea, @takehiromouri! I'll work on that.